src/backend/utils_internal_cpu/Fill_cpu.hpp

```diff
@@ -1,29 +1,39 @@
-#ifndef _H_Fill_cpu_
-#define _H_Fill_cpu_
+#ifndef SRC_BACKEND_UTILS_INTERNAL_CPU_FILL_CPU_H_
+#define SRC_BACKEND_UTILS_INTERNAL_CPU_FILL_CPU_H_
 
-#include <cstdio>
-#include <cstdlib>
-#include <stdint.h>
-#include <climits>
 #include "Type.hpp"
-#include "backend/Storage.hpp"
-#include "cytnx_error.hpp"
+
+#ifdef UNI_OMP
+  #include <omp.h>
+#endif
+
 namespace cytnx {
   namespace utils_internal {
 
-    void Fill_cpu_cd(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_cf(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_d(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_f(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_i64(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_u64(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_i32(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_u32(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_u16(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_i16(void *in, void *val, const cytnx_uint64 &Nelem);
-    void Fill_cpu_b(void *in, void *val, const cytnx_uint64 &Nelem);
+    /**
+     * @brief Assign the given value to the first `count` elements in the range beginning at
+     * `first`.
+     *
+     * This function act the same as `std::fill_n`. The execution will be parallelized when OMP is
+     * enabled.
+     *
+     * @tparam DType the data type of the elements in the range
+     *
+     * @param first the beginning of the range
+     * @param value the value to be assigned
+     * @param count the number of elements to modify
+     */
+    template <typename DType>
+    void FillCpu(void *first, const DType &value, cytnx_uint64 count) {
+      DType *typed_first = reinterpret_cast<DType *>(first);
+#ifdef UNI_OMP
+  #pragma omp parallel for schedule(dynamic)
+#endif
+      for (cytnx_uint64 i = 0; i < count; i++) {
+        *typed_first++ = value;
```

Comment:
Lets keep this dynamic here. static only give advantage when the size is known.

Reply:
I tried to decide `count` at runtim (from stdin). The static scheduling is still faster than the dynamic scheduling. And dynamic scheduling is slower than running without omp.

Reply:
for small size that is true, but it wiil get bad when it scale up. usually that can be optimize via jit but that is out of question here

Reply:
I understand that. I think what you are benchmark is when you have limited sources than your task. then in that case static is always better than dynamic. practically we usually limited this case to be just single thread because there is not much you can get from omp anyway. 

It really depends on what you are aiming for. If we are aiming for pc then yes. lets do static. then what do we gain?

Reply:
dynamic is a double blade. I doubt you can get best performance with static when you actually run on HPC.

Reply:
I dont think it matter for this PR. im ok with either of them

Reply:
My point was are we optimize for HPC or normal user, since it really doesnt matter if its 1s or 2s (even though they are 2x more) than 1 day vs 2 days. its always an crossing point for the config

Reply:
Yes, dynamic scheduling is only useful when the execution time of each iteration in the loop will vary, and you want to avoid the situation where some threads have finished their work, but there is one thread still running (and therefore blocking everything else).  But what kills the performance of the dynamic scheduling benchmark is that the default chunk size (the number of elements of the array processed at a time) is 1, for dynamic scheduling.  So each thread is only processing one element at a time.  If you specify a sensible chunk size, say 1000000, then dynamic scheduling will be basically indistinguishable from static scheduling here.  But of course there is no reason to use dynamic scheduling anyway.

I am slightly surprised that you see a benefit from OpenMP here at all - setting an array of memory to some value is going to be purely memory bandwidth limited.  A single thread will easily saturate the memory bandwidth on a modern CPU.  I modified the benchmark to add a non-OpenMP version - I also needed the `DoNotOptimize()` function because at -O3 the compiler was completely optimizing away the loop!  This shows that if the array size is small enough to fit inside the L3 cache (about 16 MB typically), then OpenMP tops out at a bit under a factor 2 faster, so probably useful.  But that is not the common case - it depends on having the array already in the cache.  For array sizes larger than the L3 cache, there is about 30% improvement with 4 threads - using more gets slower again, presumably the threads are fighting each other for memory bandwidth.  I think an optimized single-thread copy, that uses SIMD instructions, would be best here.  Surprisingly, I wasn't able to get gcc to make a SIMD-vectorized loop, not sure what I was doing wrong.

In summary, static omp with a small number of threads is slightly useful, but I suspect a SIMD optimized version would be better still.  I can't imagine any situation where the benefit from using openmp here is noticable in a realistic code.  Initializing memory to 0 is a special case that is worth treating separately, and `memset` will likely be faster than any openmp solution.

The benchmarks:
| Array size | Threads | CPU (ms) | omp-static (ms) | omp-dynamic (ms) |
| -- | -- | -- | -- | -- |
| 10'000'000 (< L3 cache) | 1 | 2.40328 | 2.41689 | 2.38274 |
| 10000000         | 2 | 2.53097 | 2.06814 | 2.05014  |
| 10000000         | 4 | 2.40613 | 1.67919 | 1.59296 |
| 100'000'000 (> L3 cache) | 1 | 32.6479 | 31.8224 | 33.9212 |
| 100'000'000  | 2 | 33.2148 | 29.2628 | 28.9037 |
| 100'000'000  | 4 | 33.7245 | 21.7398 | 22.4597 |

The code:
```c++
// compile command: g++ -std=c++17 -fopenmp -O3 -o fill fill.cpp && ./fill 1000000 4

#include <iostream>
#include <chrono>
#include <cstdlib>
#include <omp.h>

using namespace std;

// from https://github.com/google/benchmark/blob/62a321d6dc377e0ba9c712b6a8d64360616de564/include/benchmark/benchmark.h#L525
template <typename Tp>
inline void DoNotOptimize(Tp const& value) {
  asm volatile("" : : "r,m"(value) : "memory");
}


template <typename DType>
void FillCpu(void *first, DType value, size_t count) {
  DType *typed_first = reinterpret_cast<DType *>(first);
  for (int i = 0; i < count; ++i) {
    typed_first[i] = value;
  }
}

template <typename DType>
void FillCpuStatic(void *first, const DType &value, size_t count) {
   DType *typed_first = reinterpret_cast<DType *>(first);
   #pragma omp parallel for schedule(static)
   for (int i = 0; i < count; ++i) {
      typed_first[i] = value;
   }
}

template <typename DType>
void FillCpuDynamic(void *first, const DType &value, size_t count) {
  DType *typed_first = reinterpret_cast<DType *>(first);
#pragma omp parallel for schedule(dynamic,1000000)
  for (int i = 0; i < count; ++i) {
    typed_first[i] = value;
  }
}

int main(int argc, char** argv) {
  if (argc < 3) {
    cout << "expected: fill <count> <threads>\n";
    return 1;
  }
  int count = atoi(argv[1]);
  int nthreads = atoi(argv[2]);
  omp_set_num_threads(nthreads);
  int num_iterations = 10'000'000'000ll / count;
  cout << "Using " << num_iterations << " iterations." << endl;
  int *ptr = reinterpret_cast<int *>(malloc(sizeof(int) * count));
  int value = 10;

  {
    auto start = chrono::high_resolution_clock::now();
    for (int iter = 0; iter < num_iterations; ++iter) {
      FillCpu(reinterpret_cast<void *>(ptr), value, count);
      DoNotOptimize(ptr[0]);
    }
    auto end = chrono::high_resolution_clock::now();
    const std::chrono::duration<double> total_time = end - start;
    cout << "Total time for FillCpu: " << (total_time.count()*1000/num_iterations) << " milliseconds" << endl;
  }

  {
    auto start = chrono::high_resolution_clock::now();
    for (int iter = 0; iter < num_iterations; ++iter) {
      FillCpuStatic(reinterpret_cast<void *>(ptr), value, count);
      DoNotOptimize(ptr[0]);
    }
    auto end = chrono::high_resolution_clock::now();
    const std::chrono::duration<double> total_time = end - start;
    cout << "Total time for FillCpuStatic: " << (total_time.count()*1000/num_iterations) << " milliseconds" << endl;
  }

  {
    auto start = chrono::high_resolution_clock::now();
    for (int iter = 0; iter < num_iterations; ++iter) {
      FillCpuDynamic(reinterpret_cast<void *>(ptr), value, count);
      DoNotOptimize(ptr[0]);
    }
    auto end = chrono::high_resolution_clock::now();
    const std::chrono::duration<double> total_time = end - start;
    cout << "Total time for FillCpuDynamic: " << (total_time.count()*1000/num_iterations) << " milliseconds" << endl;
  }
}
```